### PR TITLE
*: Support SUBSTRING_INDEX function

### PR DESCRIPTION
**Pull request by shenli**

### CODE REVIEW

File: expression/substring_test.go [hunk 1]
```diff
@@ -87,3 +87,40 @@ func (s *testSubstringSuite) TestSubstring(c *C) {
...
 }
+
+func (s *testSubstringSuite) TestSubstringIndex(c *C) {
+	tbl := []struct {
+		str    string
+		delim  string
+		count  int64
+		result string
+	}{
+		{"www.mysql.com", ".", 2, "www.mysql"},
```
Comment (ngaut) [inline expression/substring_test.go]: Any invalid input tests?
{"www.mysql.com", ".", 20, "www.mysql"}
{"www.mysql.com", ".", -20, "www.mysql"}

Comment (shenli) [inline expression/substring_test.go]: PTAL

File: expression/substring.go [hunk 2]
```diff
@@ -112,3 +118,86 @@ func (f *FunctionSubstring) Eval(ctx context.Context, args map[interface{}]inter
...
+}
+
+// IsStatic implements the Expression IsStatic interface.
+func (f *FunctionSubstringIndex) IsStatic() bool {
+	return f.StrExpr.IsStatic() && f.Delim.IsStatic() && f.Count.IsStatic()
+}
+
+// String implements the Expression String interface.
+func (f *FunctionSubstringIndex) String() string {
+	return fmt.Sprintf("SUBSTRING_INDEX(%s, %s, %s)", f.StrExpr.String(), f.Delim.String(), f.Count.String())
```
Comment (siddontang) [inline expression/substring.go]: don't need writing String() here. `%s` can handle it automatically.

File: expression/substring.go [hunk 3]
```diff
@@ -112,3 +118,86 @@ func (f *FunctionSubstring) Eval(ctx context.Context, args map[interface{}]inter
...
+	delim, ok := t.(string)
+	if !ok {
+		return nil, errors.Errorf("Substring_Index invalid delim, need string but get %T", t)
+	}
+
+	t, err = f.Count.Eval(ctx, args)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	c, ok := t.(int64)
```
Comment (siddontang) [inline expression/substring.go]: types.ToInt64 may be safer?
Maybe return is uint64?

File: expression/substring.go [hunk 4]
```diff
@@ -112,3 +118,86 @@ func (f *FunctionSubstring) Eval(ctx context.Context, args map[interface{}]inter
...
+			end = count
+		}
+	} else {
+		// If count is negative, everything to the right of the final delimiter (counting from the right) is returned.
+		count = -count
+		if count < end {
+			start = end - count
+		}
+	}
+	substrs := strs[start:end]
```
Comment (siddontang) [inline expression/substring.go]: dont' check start and end over boundary here?

Comment (shenli) [inline expression/substring.go]: start init to 0 and will not be greater than len(strs) or small than 0
end init to len(strs) and will not be greater than len(strs) or smaller than 0

And start will be aways less or equal than end.

Comment (siddontang) [inline expression/substring.go]: got it

File: expression/substring.go (hunk 3 above)
Comment (shenli) [inline expression/substring.go]: PTAL

**Comment by siddontang** [comment]
LGTM

**Pull request merged**